sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -395,83 +482,354 @@ static struct codec_info codec_info_list[] = {
+	int i, j = 0, idx;
+	int ret;
+
+	codec = devm_kcalloc(dev, link->num_adr,
+			     sizeof(struct snd_soc_dai_link_component),
+			     GFP_KERNEL);
+	if (!codec)
+		return -ENOMEM;
+
+	ret = create_codec_dai_name(dev, link, codec);
+	if (ret < 0)
+		return ret;
+
+	part_id = SDW_PART_ID(link->adr[0]);
+	idx = find_codec_info(part_id);
+	if (idx < 0)
+		return idx;
+
+	link_id = ffs(link->mask) - 1;
+
+	/* playback & capture */
+	for (i = 0; i < 2; i++) {
+		char *name, *cpu_name;
+
+		if (!codec_info_list[idx].direction[i])
+			continue;
+
+		name = devm_kasprintf(dev, GFP_KERNEL,
+				      sdw_stream_name[i], link_id);
+		if (!name)
+			return -ENOMEM;
+
+		cpu_name = devm_kasprintf(dev, GFP_KERNEL,
+					  "SDW%d Pin%d", link_id, j + 2);
+		if (!cpu_name)
+			return -ENOMEM;
+
+		init_dai_link(dai_links, id + j, name, cpus, cpu_name,
+			      codec, link->num_adr, 1 - i, i);
+		set_codec_init_func(link, dai_links + id + j, 1 - i);
+		j++;
+	}
+
+	return j;
+}
+
+static int sof_card_dai_links_create(struct device *dev,
+				     struct snd_soc_acpi_mach *mach,
+				     struct snd_soc_card *card)
+{
+	int ssp_num = 0, sdw_num, hdmi_num = 0, dmic_num = 0;
+	struct snd_soc_dai_link_component *idisp_components;
+	struct snd_soc_acpi_mach_params *mach_params;
+	const struct snd_soc_acpi_link_adr *adr_link;
+	struct snd_soc_dai_link_component *cpus;
+	struct snd_soc_dai_link *links;
+	int num_links;
+	int i, id = 0;
+	int val;
+
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_COMMON_HDMI_CODEC)
+	hdmi_num = sof_rt711_rt1308_rt715_quirk & SOF_SDW_TGL_HDMI ?
+				SOF_TGL_HDMI_COUNT : SOF_PRE_TGL_HDMI_COUNT;
 #endif
-	{
-		.name = "SDW2-Playback",
-		.id = 3,
-		.init = second_spk_init,
-		.no_pcm = 1,
-		.dpcm_playback = 1,
-		.nonatomic = true,
-		SND_SOC_DAILINK_REG(sdw2_pin2, sdw2_codec, platform),
-	},
-};
+
+	/* check ACPI id for I2S codec */
+	if (mach->id[0] && acpi_dev_present(mach->id, NULL, -1))
+		ssp_num = 1;
+
+	mach_params = &mach->mach_params;
+	sdw_num = get_sdw_dai_link_num(mach_params);
+
+	/* enable dmic01 & dmic16k now */
+	if (sof_rt711_rt1308_rt715_quirk & SOF_SDW_PCH_DMIC)
+		dmic_num = 2;
+
+	dev_dbg(dev, "sdw %d, ssp %d, dmic %d, hdmi %d", sdw_num, ssp_num,
+		dmic_num, hdmi_num);
+
+	num_links = ssp_num + sdw_num + dmic_num + hdmi_num;
+	links = devm_kcalloc(dev, num_links, sizeof(struct snd_soc_dai_link),
+			     GFP_KERNEL);
+	cpus = devm_kcalloc(dev, num_links,
+			    sizeof(struct snd_soc_dai_link_component),
+			    GFP_KERNEL);
+	if (!links || !cpus)
+		return -ENOMEM;
+
+	/* SDW */
+	if (!sdw_num)
+		goto SSP;
+
+	adr_link = mach_params->links;
+	if (!adr_link)
+		return -EINVAL;
+
+	while (adr_link->num_adr) {
+		val = create_sdw_codec_dai(dev, links, id, cpus, adr_link);
+		if (val < 0) {
+			dev_err(dev, "failed to create dai link %d", id);
+			return -ENOMEM;
+		}
+
+		id += val;
+		adr_link++;
+	}
+
+SSP:
+	/* SSP */
+	if (ssp_num > 0) {
+		init_dai_link(links, id, "SSP2-Codec", cpus, "SSP2 Pin",
```

Comment:
can we make the SSP number part of the machine description? I don't like this hard-coding, sorry